br/pkg/lightning/errormanager/errormanager.go

```diff
@@ -504,8 +520,144 @@ func (em *ErrorManager) ResolveAllConflictKeys(
 	return errors.Trace(g.Wait())
 }
 
+// ReplaceConflictKeys query all conflicting rows (handle and their
+// values) from the current error report and resolve them
+// by replacing the necessary rows and reserving the others.
+func (em *ErrorManager) ReplaceConflictKeys(
+	ctx context.Context,
+	tbl tidbtbl.Table,
+	tableName string,
+	pool *utils.WorkerPool,
+	decoder *kv.TableKVDecoder,
+	fnGetLatest func(ctx context.Context, key []byte) ([]byte, error),
+	fnDeleteKey func(ctx context.Context, key []byte) error,
+) error {
+	if em.db == nil {
+		return nil
+	}
+
+	sessionOpts := encode.SessionOptions{
+		// TODO: need to find the correct value for SQLMode
+		SQLMode: mysql.ModeStrictAllTables,
+	}
+	encoder, err := kv.NewBaseKVEncoder(&encode.EncodingConfig{
+		Table:          tbl,
+		SessionOptions: sessionOpts,
+		Logger:         em.logger,
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	g, gCtx := errgroup.WithContext(ctx)
+	pool.ApplyOnErrorGroup(g, func() error {
+		// TODO: provide a detailed document to explain the algorithm and link it here
+		// demo for "replace" algorithm: https://github.com/lyzx2001/tidb-conflict-replace
+		// check index KV first
+		rawKeyRows, err := em.db.QueryContext(
+			gCtx, fmt.Sprintf(selectIndexConflictKeysReplace, em.schemaEscaped),
+			tableName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for rawKeyRows.Next() {
+			var rawKey, rawValue, rawHandle []byte
+			var indexName string
+			if err := rawKeyRows.Scan(&rawKey, &indexName, &rawValue, &rawHandle); err != nil {
+				return errors.Trace(err)
+			}
+			em.logger.Debug("got raw_key, index_name, raw_value, raw_handle from table",
+				zap.Binary("raw_key", rawKey),
+				zap.String("index_name", indexName),
+				zap.Binary("raw_value", rawValue),
+				zap.Binary("raw_handle", rawHandle))
+
+			// get the latest value of rawKey from downstream TiDB
+			value, err := fnGetLatest(gCtx, rawKey)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			em.logger.Debug("got latest value from fnGetLatest",
+				zap.Binary("raw_key", rawKey),
+				zap.Binary("latest value", value))
+
+			// if the latest value of rawKey equals to rawValue, that means this index KV is maintained in downstream TiDB
+			// if not, that means this index KV has been overwrt, and its corresponding data KV needs to be deleted
+			if bytes.Equal(rawValue, value) {
+				continue
+			}
+			indexInfo := tbl.Meta().FindIndexByName(indexName)
+			handle, err := decoder.DecodeHandleFromIndex(indexInfo, rawKey, rawValue)
+			em.logger.Debug("got handle from index",
+				zap.String("handle", handle.String()),
+				zap.Error(err))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			// encode the row key of the data KV that needs to be deleted
+			rowKey := tablecodec.EncodeRowKeyWithHandle(tbl.Meta().ID, handle)
```

Comment:
can we get rowkey using `EncodeRowKey` with `rawHandle`

Reply:
I have tried to replace `rowKey := tablecodec.EncodeRowKeyWithHandle(tbl.Meta().ID, handle)` with `rowKey := tablecodec.EncodeRowKey(tbl.Meta().ID, rawHandle)`, and it turns out the two returned `rowKey` are different:
```
rowKey from EncodeRowKeyWithHandle: 74800000000000004b5f728000000000000002
rowKey from EncodeRowKey: 74800000000000004b5f7274800000000000004b5f728000000000000002
```
If we use the result from `tablecodec.EncodeRowKey`, the unit test and integration test would fail, so we should just keep the current `tablecodec.EncodeRowKeyWithHandle`.

Reply:
@lyzx2001 Seems `rawHandle` is `rowKey`

Reply:
see https://github.com/pingcap/tidb/pull/45926#discussion_r1295841154

Reply:
> @lyzx2001 Seems `rawHandle` is `rowKey`

Thanks for reminding. The encoding for `rowKey` has been deleted, and I directly use `rawHandle` as the row key of the data KV that needs to be deleted.

Reply:
> You still use encoded `rowKey` in line 598?

In the newest commit I use `rawHandle` here